### PR TITLE
Add cycle reporter to handle the case where a non-toolchain is passed…

### LINKED ISSUE #7171: java.lang.IllegalStateException when referencing ToolchainInfo targets instead of ToolchainRuleInfo
Bazel's toolchain registration involves two targets: a "toolchain info" that returns a `platform_common.ToolchainInfo` with arbitrary content, and a "toolchain" that registers the toolchain info with execution constraints.

If the user confuses these two (either when calling `native.register_toolchains()` or with the `--extra_toolchains` flag) then Bazel will crash during analysis.

example BUILD ([`rules_m4/m4/toolchains/BUILD`](https://github.com/jmillikin/rules_m4/blob/5a8562a8796963da93b6ed8e6c205551cbf949b5/m4/toolchains/BUILD)):
```
load("//m4:m4.bzl", "M4_VERSIONS")
load("//m4:toolchain.bzl", "m4_toolchain_info")

[m4_toolchain_info(
    name = "v{}".format(version),
    m4 = "@m4_v{}//bin:m4".format(version),
    tags = ["manual"],
    visibility = ["//visibility:public"],
) for version in M4_VERSIONS]

[toolchain(
    name = "v{}_toolchain".format(version),
    toolchain = ":v{}".format(version),
    toolchain_type = "//m4:toolchain_type",
) for version in M4_VERSIONS]
```

Full error output when the user forgets the `_toolchain` suffix:
```
$ ./bazel-0.22.0rc1-darwin-x86_64 build //tests:expansion_test --extra_toolchains=@io_bazel_rules_m4//m4/toolchains:v1.4.18
Starting local Bazel server and connecting to it...
INFO: Invocation ID: 2780f5f3-5e47-4588-805e-0b2ce38fc86e
Internal error thrown during build. Printing stack trace: java.lang.IllegalStateException: topLevelKey: //tests:expansion_test BuildConfigurationValue.Key[94f032dd1c40f99180ff553125967cf3] false
alreadyReported: false
path to cycle:
//tests:expansion_test BuildConfigurationValue.Key[94f032dd1c40f99180ff553125967cf3] false
Key{configurationKey=BuildConfigurationValue.Key[94f032dd1c40f99180ff553125967cf3], toolchainTypeLabel=@io_bazel_rules_m4//m4:toolchain_type, targetPlatformKey=@bazel_tools//platforms:target_platform BuildConfigurationValue.Key[94f032dd1c40f99180ff553125967cf3] false, availableExecutionPlatformKeys=[@bazel_tools//platforms:host_platform BuildConfigurationValue.Key[94f032dd1c40f99180ff553125967cf3] false]}
RegisteredToolchainsValue.Key{configurationKey: BuildConfigurationValue.Key[94f032dd1c40f99180ff553125967cf3]}
@io_bazel_rules_m4//m4/toolchains:v1.4.18 BuildConfigurationValue.Key[94f032dd1c40f99180ff553125967cf3] false
cycle:
@io_bazel_rules_m4//m4/internal:deny_shell BuildConfigurationValue.Key[a1782a45cb1a1c8fbec9cef9245c752c] true
Key{configurationKey=BuildConfigurationValue.Key[a1782a45cb1a1c8fbec9cef9245c752c], toolchainTypeLabel=@bazel_tools//tools/cpp:toolchain_type, targetPlatformKey=@bazel_tools//platforms:host_platform BuildConfigurationValue.Key[a1782a45cb1a1c8fbec9cef9245c752c] true, availableExecutionPlatformKeys=[@bazel_tools//platforms:host_platform BuildConfigurationValue.Key[a1782a45cb1a1c8fbec9cef9245c752c] true]}
RegisteredToolchainsValue.Key{configurationKey: BuildConfigurationValue.Key[a1782a45cb1a1c8fbec9cef9245c752c]}
@io_bazel_rules_m4//m4/toolchains:v1.4.18 BuildConfigurationValue.Key[a1782a45cb1a1c8fbec9cef9245c752c] true

	at com.google.common.base.Preconditions.checkState(Preconditions.java:507)
	at com.google.devtools.build.skyframe.CyclesReporter.reportCycles(CyclesReporter.java:85)
	at com.google.devtools.build.lib.skyframe.SkyframeBuildView.processErrors(SkyframeBuildView.java:527)
	at com.google.devtools.build.lib.skyframe.SkyframeBuildView.configureTargets(SkyframeBuildView.java:444)
	at com.google.devtools.build.lib.analysis.BuildView.update(BuildView.java:369)
	at com.google.devtools.build.lib.buildtool.AnalysisPhaseRunner.runAnalysisPhase(AnalysisPhaseRunner.java:209)
	at com.google.devtools.build.lib.buildtool.AnalysisPhaseRunner.execute(AnalysisPhaseRunner.java:118)
	at com.google.devtools.build.lib.buildtool.BuildTool.buildTargets(BuildTool.java:143)
	at com.google.devtools.build.lib.buildtool.BuildTool.processRequest(BuildTool.java:253)
	at com.google.devtools.build.lib.runtime.commands.BuildCommand.exec(BuildCommand.java:83)
	at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.execExclusively(BlazeCommandDispatcher.java:477)
	at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.exec(BlazeCommandDispatcher.java:205)
	at com.google.devtools.build.lib.server.GrpcServerImpl.executeCommand(GrpcServerImpl.java:749)
	at com.google.devtools.build.lib.server.GrpcServerImpl.access$1600(GrpcServerImpl.java:103)
	at com.google.devtools.build.lib.server.GrpcServerImpl$2.lambda$run$0(GrpcServerImpl.java:818)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)

INFO: Elapsed time: 4.065s
```

### PR DESCRIPTION
**Pull request by katre**
… to register_toolchains.

Fixes #7171.

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/skyframe/RegisteredToolchainsCycleReporter.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  @Override
+  public boolean maybeReportCycle(
+      SkyKey topLevelKey,
+      CycleInfo cycleInfo,
+      boolean alreadyReported,
+      ExtendedEventHandler eventHandler) {
+    ImmutableList<SkyKey> pathToCycle = cycleInfo.getPathToCycle();
+    ImmutableList<SkyKey> cycle = cycleInfo.getCycle();
+    if (pathToCycle.isEmpty()) {
+      return false;
```
Comment (haxorz) [inline src/main/java/com/google/devtools/build/lib/skyframe/RegisteredToolchainsCycleReporter.java]: why is this? is it b/c we don't expect the SkyKeys of relevance to be top-level keys of a skyframe evaluation?

please add a comment.

File: src/test/shell/bazel/toolchain_test.sh [hunk 2]
```diff
@@ -1076,6 +1076,117 @@ function test_local_config_platform() {
   expect_log '@local_config_platform//:host'
 }
 
+# Test cycles in registered toolchain dtaa, which can only happen when
```
Comment (haxorz) [inline src/test/shell/bazel/toolchain_test.sh]: > "dtaa"

not sure if type ("data"?) or bad/copy paste ("... registered_toolchains").

File: src/main/java/com/google/devtools/build/lib/skyframe/RegisteredToolchainsCycleReporter.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    Function<SkyKey, String> printer =
+        new Function<SkyKey, String>() {
+          @Override
+          public String apply(SkyKey input) {
+            if (input.argument() instanceof ConfiguredTargetKey) {
+              Label label = ((ConfiguredTargetKey) input.argument()).getLabel();
+              return label.toString();
+            }
+            if (input.argument() instanceof RegisteredToolchainsValue.Key) {
+              return "All registered toolchains";
```
Comment (haxorz) [inline src/main/java/com/google/devtools/build/lib/skyframe/RegisteredToolchainsCycleReporter.java]: > "All registered toolchains"

you're the expert on that RegisteredToolchainsValue.Key, not me. but this SkyKey -> String transformation seems lossy. at the very least, the word "all" seems misleading.

File: src/main/java/com/google/devtools/build/lib/skyframe/RegisteredToolchainsCycleReporter.java (hunk 1 above)
Comment (katre) [reply src/main/java/com/google/devtools/build/lib/skyframe/RegisteredToolchainsCycleReporter.java]: I copied it from the skylark cycle reporter, but actually this shouldn't care, so I've removed the check.

File: src/test/shell/bazel/toolchain_test.sh (hunk 2 above)
Comment (katre) [reply src/test/shell/bazel/toolchain_test.sh]: Fixed the comment, thanks for catching that.

File: src/main/java/com/google/devtools/build/lib/skyframe/RegisteredToolchainsCycleReporter.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+      return false;
+    } else if (alreadyReported) {
+      return true;
+    } else if (!Iterables.any(cycle, IS_REGISTERED_TOOLCHAINS_SKY_KEY)
+        || !Iterables.any(cycle, IS_CONFIGURED_TARGET_SKY_KEY)
+        || !Iterables.any(cycle, IS_TOOLCHAIN_RESOLUTION_SKY_KEY)) {
+      return false;
+    }
+
+    // Find the ConfiguredTargetKey, this should tell the problem.
```
Comment (haxorz) [inline src/main/java/com/google/devtools/build/lib/skyframe/RegisteredToolchainsCycleReporter.java]: perhaps a nit:

> ... the ConfiguredTargetKey
> ... .findFirst()

note that the order of the cycle is out of the control of this class (depends on blaze-on-skyframe parallelism, the top level keys requests, cached state, etc). do you perhaps mean "Find *a* ConfiguredTargetKey"? or is the cycle expected to contain exactly one ConfiguredTargetKey?

File: src/main/java/com/google/devtools/build/lib/skyframe/RegisteredToolchainsCycleReporter.java (hunk 3 above)
Comment (katre) [reply src/main/java/com/google/devtools/build/lib/skyframe/RegisteredToolchainsCycleReporter.java]: I've changed it. I didn't want to make it too skyframe-dependent, but you're right, in the cycle list, it's a bit confusing.

File: src/main/java/com/google/devtools/build/lib/skyframe/RegisteredToolchainsCycleReporter.java (hunk 1 above)
Comment (haxorz) [reply src/main/java/com/google/devtools/build/lib/skyframe/RegisteredToolchainsCycleReporter.java]: yes, i noticed that. i'd have the same comment about the skylark cycle reporter code :p

File: src/main/java/com/google/devtools/build/lib/skyframe/RegisteredToolchainsCycleReporter.java (hunk 4 above)
Comment (katre) [reply src/main/java/com/google/devtools/build/lib/skyframe/RegisteredToolchainsCycleReporter.java]: There may be cases where there are multiple configured targets, but the correct one to display is the one that is the most direct dependency of RegisteredToolchainFunction.

**Comment by haxorz** [comment]
> Sample of the error message:
> 
> ```
> ERROR: Misconfigured toolchains: //:upper_toolchain is declared as a toolchain but has inappropriate dependencies. Declared toolchains should be created with the 'toolchain' rule and should not have dependencies that themselves require toolchains.
> .-> All registered toolchains
> |   //:upper_toolchain
> |   toolchain type //:lower
> `-- All registered toolchains
> WARNING: errors encountered while analyzing target '//:upper_lib': it will not be built
> ```
> Is this clear enough to explain the problem? @jmillikin what do you think?

maybe use the word "cycle" somewhere to signify to the user that what they're doing is really bad.

perhaps "... but has inappropriate cyclic dependencies" or "has inappropriate dependencies that cause a cycle inside of Bazel".

up to you.

File: src/main/java/com/google/devtools/build/lib/skyframe/RegisteredToolchainsCycleReporter.java (hunk 4 above)
Comment (haxorz) [reply src/main/java/com/google/devtools/build/lib/skyframe/RegisteredToolchainsCycleReporter.java]: > but the correct one to display is the one that is the most direct dependency of RegisteredToolchainFunction.

so are you going to change the code to do that then?

**Comment by jmillikin** [comment]
That error message is a good start but I found it a bit too long to be useful. I would put the non-fatal part (unexpected rule type) in a warning, so it's clear that there are two distinct problems:

> WARNING: Registered toolchain //:upper_toolchain was not declared with the 'toolchain' rule.
>
> ERROR: Registered toolchain //:upper_toolchain must not have dependencies that themselves require toolchains.

**Comment by katre** [comment]
Leaving the error message as-is and starting the import process. Feel free to file additional issues with suggestions for improvements to the message.

**Pull request closed without merging**